extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java

```diff
@@ -620,17 +629,20 @@ public void runWith(Runnable task, Object context) {
                 // Only do context handling if it's non-null
                 if (context != null && context != currentContext) {
                     ContextInternal vertxContext = (ContextInternal) context;
-                    // The CDI contexts must not be propagated
-                    // First test if VertxCurrentContextFactory is actually used
-                    if (currentContextFactory != null) {
-                        List<String> keys = currentContextFactory.keys();
+                    if (!ignoredKeysSuppliers.isEmpty()) {
                         ConcurrentMap<Object, Object> local = vertxContext.localContextData();
-                        if (containsScopeKey(keys, local)) {
-                            // Duplicate the context, copy the data, remove the request context
-                            vertxContext = vertxContext.duplicate();
-                            vertxContext.localContextData().putAll(local);
-                            keys.forEach(vertxContext.localContextData()::remove);
-                            VertxContextSafetyToggle.setContextSafe(vertxContext, true);
+                        ContextInternal duplicateVertxContext = null;
+                        for (Supplier<List<String>> ignoredKeysSupplier : ignoredKeysSuppliers) {
+                            List<String> keys = ignoredKeysSupplier.get();
+                            if (containsScopeKey(keys, local)) {
+                                if (duplicateVertxContext == null) {
+                                    // Duplicate the context, copy the data, remove the request context
+                                    duplicateVertxContext = vertxContext = vertxContext.duplicate();
```

Comment:
I think that we should only make this once, i.e. when the local context contains any of the ignored keys...

Reply:
this is done once, I am not sure I understand what you mean

Reply:
I'll change it based on the fact that we switch from supplier to the list so it will be same code as before. I think it was executed only once, maybe you meant something else than I do.

Reply:
Well, you iterate over all suppliers and if a key from a specific supplier is found then a new duplicated context is created. Which means that if you have two suppliers and both provide a key that is contained in the local context then two duplicated contexts will be created. Am I missing something?

Reply:
`ContextInternal duplicateVertxContext = null;` is in for loop outer scope, it doesn't change with each iteration, once the context is duplicated, the `duplicateVertxContext == null` condition is false.